src/sha256.nr

```diff
@@ -512,6 +485,150 @@ fn hash_final_block(msg_block: MSG_BLOCK, mut state: STATE) -> HASH {
+        }
+    }
+
+    // If we had modulo == 0 then it means the last block was full,
+    // and we can reset the pointer to zero to overwrite it.
+    if msg_byte_ptr == BLOCK_SIZE {
+        msg_byte_ptr = 0;
+    }
+
+    // Pad the rest such that we have a [u32; 2] block at the end representing the length
+    // of the message, and a block of 1 0 ... 0 following the message (i.e. [1 << 7, 0, ..., 0]).
+    // Here we rely on the fact that everything beyond the available input is set to 0.
+    let index = msg_byte_ptr / INT_SIZE;
+    msg_block[index] = set_item_byte_then_zeros(msg_block[index], msg_byte_ptr, 1 << 7);
+
+    msg_byte_ptr = msg_byte_ptr + 1;
+    let last_block = msg_block;
+
+    // If we don't have room to write the size, compress the block and reset it.
+    if msg_byte_ptr > MSG_SIZE_PTR {
+        h = sha256_compression(msg_block, h);
+
+        // `attach_len_to_msg_block` will zero out everything after the `msg_byte_ptr`.
+        msg_byte_ptr = 0;
+    }
+
+    msg_block = unsafe { attach_len_to_msg_block(msg_block, msg_byte_ptr, message_size) };
+
+    verify_msg_len(msg_block, last_block, msg_byte_ptr, message_size);
+
+    hash_final_block(msg_block, h)
+}
+
+/**
+ * Given some state of a partially computed sha256 hash and part of the preimage, continue hashing
+ * @notice used for complex/ recursive offloading of post-partial hashing
+ *
+ * @param N - the maximum length of the message to hash
+ * @param h - the intermediate hash state
+ * @param msg - the preimage to hash
+ * @param message_size - the actual length of the preimage to hash
+ * @return the intermediate hash state after compressing in msg to h
+ */
+pub fn partial_sha256_var_interstitial<let N: u32>(
+    mut h: [u32; 8],
+    msg: [u8; N],
+    message_size: u32,
+) -> [u32; 8] {
+    assert(message_size % BLOCK_SIZE == 0, "Message size must be a multiple of the block size");
+    if std::runtime::is_unconstrained() {
+        unsafe {
+            __sha_partial_var_interstitial(h, msg, message_size)
+        }
+    } else {
+        let (mut h, _, _) = process_full_blocks(msg, message_size, N, h);
+
+        h
+    }
+}
+
+/**
+ * Given some state of a partially computed sha256 hash and remaining preimage, complete the hash
+ * @notice used for traditional partial hashing
+ *
+ * @param N - the maximum length of the message to hash
+ * @param h - the intermediate hash state
+ * @param msg - the remaining preimage to hash
+ * @param message_size - the size of the remaining preimage
+ * @param real_message_size -= the total size of the preimage
+ * @return finalized sha256 hash
+ */
+pub fn partial_sha256_var_end<let N: u32>(
+    mut h: [u32; 8],
+    msg: [u8; N],
+    message_size: u32,
+    real_message_size: u32,
+) -> [u8; 32] {
+    assert(message_size % BLOCK_SIZE == 0, "Message size must be a multiple of the block size");
+    if std::runtime::is_unconstrained() {
+        unsafe {
```

Comment:
Can you add a safety comment here? We should aim to avoid adding more warnings into the library.